airflow/operators/dagrun_operator.py

```diff
@@ -18,81 +18,64 @@
-        a DagRun created. This ``obj`` object contains a ``run_id`` and
-        ``payload`` attribute that you can modify in your function.
-        The ``run_id`` should be a unique identifier for that DAG run, and
-        the payload has to be a picklable object that will be made available
-        to your tasks while executing that DAG run. Your function header
-        should look like ``def foo(context, dag_run_obj):``
-    :type python_callable: python callable
+    :param conf: Configuration for the DAG run
+    :type conf: dict
     :param execution_date: Execution date for the dag (templated)
     :type execution_date: str or datetime.datetime
     """
-    template_fields = ('trigger_dag_id', 'execution_date')
-    ui_color = '#ffefeb'
+
+    template_fields = ("trigger_dag_id", "execution_date", "conf")
+    ui_color = "#ffefeb"
 
     @apply_defaults
     def __init__(
-            self,
-            trigger_dag_id: str,
-            python_callable: Optional[Callable[[Dict, DagRunOrder], DagRunOrder]] = None,
-            execution_date: Optional[Union[str, datetime.datetime]] = None,
-            *args, **kwargs) -> None:
+        self,
+        trigger_dag_id: str,
+        conf: Optional[Dict] = None,
+        execution_date: Optional[Union[str, datetime.datetime]] = None,
+        *args,
+        **kwargs
+    ) -> None:
         super().__init__(*args, **kwargs)
-        self.python_callable = python_callable
         self.trigger_dag_id = trigger_dag_id
+        self.conf = conf
 
-        self.execution_date = None  # type: Optional[Union[str, datetime.datetime]]
-        if isinstance(execution_date, datetime.datetime):
-            self.execution_date = execution_date.isoformat()
-        elif isinstance(execution_date, str):
+        if execution_date is None or isinstance(execution_date, (str, datetime.datetime)):
             self.execution_date = execution_date
-        elif execution_date is None:
-            self.execution_date = None
         else:
             raise TypeError(
-                'Expected str or datetime.datetime type '
-                'for execution_date. Got {}'.format(
-                    type(execution_date)))
+                "Expected str or datetime.datetime type for execution_date. "
+                "Got {}".format(type(execution_date))
+            )
 
-    def execute(self, context):
-        if self.execution_date is not None:
-            run_id = 'trig__{}'.format(self.execution_date)
-            self.execution_date = timezone.parse(self.execution_date)
+    def execute(self, context: Dict):
+        if isinstance(self.execution_date, datetime.datetime):
+            run_id = "trig__{}".format(self.execution_date.isoformat())
+        elif isinstance(self.execution_date, str):
+            run_id = "trig__{}".format(self.execution_date)
+            self.execution_date = timezone.parse(self.execution_date)  # trigger_dag() expects datetime
```

Comment:
Wouldn't it be better to set the execution date on line 58 where you are setting it the first time?
```python
if isinstance(execution_date, str):
    self.execution_date = timezone.parse(self.execution_date)
else:
    self.execution_date = execution_date
```
IMO `self.execution_date = timezone.parse(self.execution_date)` is a kind of validation so that should be made in the constructor even if it will be called more often than in the `execute`.

then you can just do
```python
if self.execution_date is None:
    self.execution_date = timezone.utcnow()
run_id = "trig__{}".format(self.execution_date.isoformat())
```

WDYT?

Reply:
Much simpler, much better👍

Reply:
Would like to use it as it is much shorter and simpler.

However, the `execution_date` can be templated. For example `{{ execution_date }}` will fail in `timezone.parse()`. So, we have to save it first, wait for `execute()` to be called and all variables to be templated, and only then can we call `timezone.parse()` on the `execution_date` :(